### PR TITLE
4 train relevance classifier

### LINKED ISSUE #4: Train relevance classifier
### Problem to solve

As `scientist`, I want `have a classifier that predicts the relevance of Tweets` so that `I can only consider "relevant" Tweets for my rain classification Task`.

#### Further details

Initial implementation done during maelstrom bootcamp (see [a2 repo](https://gitlab.com/4cast1/maelstrom/a2/a2/-/blob/main/bootcamp2023/Day3/05FalconDatasetbuilding.ipynb?ref_type=heads)).

"Relevance"/"relevant" Tweets are Tweets that contain sufficient information for a human/AI to determine if it was raining or not raining.

### Proposal

- [x] Bring [relevance dataset](https://github.com/rajhaq/AP2-Social-media-data-for-better-local-forecasts/issues/2) into form such that a train and test set can be created for this task
  - [x]  All Tweets that have a relevance score predicted by falcon should now get the score as a label that is either "relevant" or "not relevant"
    - [x] This depends on the relevance score threshold, we pick 0.5 for our first attempt
- [x] Adopt rain classifier to allow for classification of relevance of Tweets 
  - [x] Use you rain classifier notebook as basis for this. (make a copy) 
  - [x] Evaluate the performance of your model.
    - [x] Check manually if predictions of model make sense
      - [x] For easy and more difficult (especially misclassified) Tweets
- [x] Change relevance score threshold and see how it affects performance

### What does success look like, and how can we measure that?

- [ ] Trained relevance classifier exists 
  - [ ]  performance is estimated
  - [ ] performance based on relevance score threshold is estimated

- weight: 5

### PR DESCRIPTION
**Pull request by rajhaq**

### CODE REVIEW

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 1]
```diff
@@ -0,0 +1,283 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Train our Baseline Model with New Dataset (DeBERTa)\n",
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]:
```suggestion
    "# Train our Baseline Relevance Classifier Model with New Dataset (DeBERTa)\n",
```

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 2]
```diff
@@ -0,0 +1,283 @@
+{
+ "cells": [
+  {
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]: mention relevance classifier in filename, otherwise difficult to distinguish rain classifier from relevance classifier

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Train our Baseline Model with New Dataset (DeBERTa)\n",
+    "\n",
+    "## Overview\n",
+    "\n",
+    "This notebook trains a classifier using the [DeBERTa small](https://huggingface.co/microsoft/deberta-v3-small/tree/main) model, applied to a new dataset generated by the Falcon 40B model from the ER5 tweet dataset. The data is formatted as a CSV file, from our methodology from the Bootcamp day 2 reference, 04ModelTrelevance.ipynb.\n",
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]:
```suggestion
    "This notebook trains a relevance classifier using the [DeBERTa small](https://huggingface.co/microsoft/deberta-v3-small/tree/main) model, applied to a new dataset generated by the Falcon 40B model from the ER5 tweet dataset. The data is formatted as a CSV file, from our methodology from the Bootcamp day 2 reference, 04ModelTrelevance.ipynb.\n",
```

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    "\n",
+    "## Key Steps and Objectives\n",
+    "\n",
+    "1. **Classifier Training**: We fine-tune the DeBERTa small model on the Falcon 40B dataset to adapt it to this specific classification task.\n",
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]:
```suggestion
    "1. **Classifier Training**: We fine-tune the DeBERTa small model on the Falcon 40B dataset to adapt it to the specific task of classifying the relevance of Tweets.\n",
```

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    "\n",
+    "2. **Results Visualization**: We analyze the model's performance through:\n",
+    "   - **Confusion Matrix**: Provides insights into accuracy and misclassification rates.\n",
+    "   - **ROC Curve**: Illustrates the model's performance across various thresholds.\n",
+    "\n",
+    "This notebook is essential for evaluating the DeBERTa-based classifier on the new dataset and offers comprehensive insights through detailed visualizations."
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]:
```suggestion
    "This notebook is essential for evaluating the DeBERTa-based relevance classifier on the new dataset and offers comprehensive insights through detailed visualizations."
```

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Again define labels\n",
+    "key_tp = \"content\"\n",
+    "ds_tweets[\"content\"] = ds_tweets[key_tp].astype(str)"
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]:
```suggestion
    "key_input = \"content\"\n",
    "ds_tweets[\"content\"] = ds_tweets[key_input].astype(str)"
```
not talking about **T**otal **P**recipitation (=tp) here, variable names should reflect this

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    "\n",
+    "parameters = {\n",
+    "    \"learning_rate\": 8e-5,\n",
+    "    \"batch_size\": 16,\n",
+    "    \"weight_decay\": 0.01,\n",
+    "    \"epochs\": 1,\n",
+    "    \"warmup_ratio\": 0.1,\n",
+    "    \"cls_dropout\": 0.3,\n",
+    "    \"lr_scheduler_type\": \"cosine\",\n",
+    "}\n",
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]: did you use the baseline parameters that are used for training the rain classifier? if so, please add a comment clarifying this in the code

File: 07deBERTaClassifierForNewDataset.ipynb [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    "## Key Steps and Objectives\n",
+    "\n",
+    "1. **Classifier Training**: We fine-tune the DeBERTa small model on the Falcon 40B dataset to adapt it to this specific classification task.\n",
+    "\n",
+    "2. **Results Visualization**: We analyze the model's performance through:\n",
+    "   - **Confusion Matrix**: Provides insights into accuracy and misclassification rates.\n",
+    "   - **ROC Curve**: Illustrates the model's performance across various thresholds.\n",
+    "\n",
+    "This notebook is essential for evaluating the DeBERTa-based classifier on the new dataset and offers comprehensive insights through detailed visualizations."
+   ]
```
Comment (kristian4cast) [inline 07deBERTaClassifierForNewDataset.ipynb]: Could you comment on Dataset generation
- Threshold above which Tweets are considered relevant:
- Link to notebook that generated the Tweets
- Ratio between relevant and not relevant Tweets:

Could you comment on performance of the model:
- Tweets in training set: ??
- Tweets in test set:
- roc: ??
- f1 relevant ??, f1 not relevant ??

This allows us to keep some record on your results....

**Review by kristian4cast — CHANGES_REQUESTED** [review]

**Review by kristian4cast — APPROVED** [review]
